src/devices/src/virtio/net/persist.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines the structures needed for saving/restoring net devices.
+
+use std::io;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+
+use dumbo::{ns::MmdsNetworkStack, persist::MmdsNetworkStackState, MAC_ADDR_LEN};
+use rate_limiter::{persist::RateLimiterState, RateLimiter};
+use snapshot::Persist;
+use versionize::{VersionMap, Versionize, VersionizeResult};
+use versionize_derive::Versionize;
+use vm_memory::GuestMemoryMmap;
+
+use super::device::Net;
+
+use crate::virtio::persist::VirtioDeviceState;
+use crate::virtio::{DeviceState, Queue};
+
+#[derive(Versionize)]
+pub struct NetConfigSpace {
+    guest_mac: [u8; MAC_ADDR_LEN],
+}
+
+#[derive(Versionize)]
+pub struct NetState {
+    id: String,
+    tap_if_name: String,
+    rx_rate_limiter_state: RateLimiterState,
+    tx_rate_limiter_state: RateLimiterState,
+    mmds_ns: Option<MmdsNetworkStackState>,
+    config_space: NetConfigSpace,
+    virtio_state: VirtioDeviceState,
+}
+
+pub struct NetConstructorArgs {
+    mem: GuestMemoryMmap,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    CreateNet(super::Error),
+    CreateRateLimiter(io::Error),
+}
+
+impl Persist for Net {
+    type State = NetState;
+    type ConstructorArgs = NetConstructorArgs;
+    type Error = Error;
+
+    fn save(&self) -> Self::State {
+        let mut guest_mac = [0; MAC_ADDR_LEN];
+        guest_mac.copy_from_slice(&self.config_space);
+
+        NetState {
+            id: self.id().clone(),
+            tap_if_name: self.tap_if_name.clone(),
+            rx_rate_limiter_state: self.rx_rate_limiter.save(),
+            tx_rate_limiter_state: self.tx_rate_limiter.save(),
+            mmds_ns: self.mmds_ns.as_ref().map(|mmds| mmds.save()),
+            config_space: NetConfigSpace { guest_mac },
+            virtio_state: VirtioDeviceState::from_device(self),
+        }
+    }
+
+    fn restore(
+        constructor_args: Self::ConstructorArgs,
+        state: &Self::State,
+    ) -> std::result::Result<Self, Self::Error> {
+        // Rate limiter restore can fail at creating a timerfd.
+        let rx_rate_limiter = RateLimiter::restore((), &state.rx_rate_limiter_state)
+            .map_err(Error::CreateRateLimiter)?;
+        let tx_rate_limiter = RateLimiter::restore((), &state.tx_rate_limiter_state)
+            .map_err(Error::CreateRateLimiter)?;
+        let mut net = Net::new_with_tap(
+            state.id.clone(),
+            state.tap_if_name.clone(),
+            None,
+            rx_rate_limiter,
+            tx_rate_limiter,
+            state.mmds_ns.is_some(),
+        )
+        .map_err(Error::CreateNet)?;
+
+        net.mmds_ns = state
+            .mmds_ns
+            .as_ref()
+            .map(|mmds_state| MmdsNetworkStack::restore((), &mmds_state).unwrap());
```

Comment:
`MmdsNetworkStack::restore()` cannot fail _now_, but if we change that code and introduce any error conditions, this `unwrap()` will most likely be overlooked and might create problems.

We should avoid unwrap()s or add code comments in the places where the unwrap-is-safe assumption originates (in this case `MmdsNetworkStack::restore`)

Reply:
Done.